File: Data Structures & Algorithms/validate-parentheses/submission-7.py
class Solution:
    def isValid(self, s: str) -> bool:
        openBracket = { '{': '}', '[': ']', '(': ')' } 

        stack = [] 

        for b in s: 
            if b in openBracket: 
                stack.append(b) 
            else:  
                if len(stack) == 0: 
                    return False 
                    
                if stack: 
                    last = stack.pop() 
                    if last not in openBracket or openBracket[last] != b: 
                        return False 

        return True if len(stack) == 0 else False

File: Data Structures & Algorithms/sliding-window-maximum/submission-7.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number[]}
     */
    maxSlidingWindow(nums, k) { 
        let stack = []
        let result = [] 
        let l = 0 
        for (let r = 0; r < nums.length; r ++) {
            let cur = nums[r]

            while(stack.length > 0 && cur > nums[stack[stack.length - 1]]) {
                stack.pop()
            }

            stack.push(r)

            if (l > stack[0]) {
                stack.shift()
            }

            if ((r - l + 1) === k) {
                result.push(nums[stack[0]])
                l ++
            }
        }

        return result
    }
}
